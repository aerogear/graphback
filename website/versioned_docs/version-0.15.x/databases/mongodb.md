---
id: mongodb
title: Using MongoDB in Graphback
sidebar_label: MongoDB 
---

[MongoDB](https://www.mongodb.com/) is a general purpose, document-based, distributed database built for modern application developers.
[Graphback MongoDB Provider](https://www.npmjs.com/package/@graphback/runtime-mongo) package provides instant integration to your Mongo database, giving you a full implementation of the [CRUD API](../crud/introduction.md). 

The package is built on top of the official [MongoDB Node.js Driver](https://mongodb.github.io/node-mongodb-native).

## Installation

Install with npm:

```bash
npm install @graphback/runtime-mongo mongodb
```

Install with yarn:

```bash
yarn add @graphback/runtime-mongo mongodb
```

Running either of the two commands will install the `Graphback MongoDB Provider` ([`@graphback/runtime-mongo`](https://www.npmjs.com/package/@graphback/runtime-mongo)) and the MongoDB Driver ([`mongodb`](https://www.npmjs.com/package/mongodb)).

## Creating Database Connection

You'll need a running MongoDB instance before initializing the connection. 

:::info
If you do not have a running instance, you can use [Docker](https://www.docker.com/) to quickly spin up a MongoDB container by following the instructions given in
[MongoDB Docker Image](https://hub.docker.com/_/mongo).
:::

The following code can be used to connect to your MongoDB.

```ts
import { Db, MongoClient } from "mongodb";

// Example Mongo Connection URL
const url = 'mongodb://localhost:27017';

const mongoClient: MongoClient = await MongoClient.connect(url);
// get db instance
const db: Db = mongoClient.db('users');
```

Here we are connecting to the `users` database, on a MongoDB instance running locally.

You can change the configurations variables to match the one you have.

To learn more about the accepted configuration, visit the [MongoDB Driver Connection Options](https://mongodb.github.io/node-mongodb-native/3.5/reference/connecting/) documentation.

## Using MongoDB Provider

The Graphback MongoDB provider exposes a [`createMongoDbProvider`](../api/graphback-runtime-mongodb/modules/_createmongodbprovider_.md) method, which can be used to create data providers for each of your  data models. 

The code below shows how you can create such a data provider creator and how it can be passed to [`buildGraphbackAPI`](./api/graphback/modules/_buildgraphbackapi_.md).

```ts
import { Db, MongoClient } from "mongodb";
import { buildGraphbackAPI } from 'graphback';
import { createMongoDbProvider } from '@graphback/runtime-mongo';

// highlight-start
// Example Mongo Connection URL
const url = 'mongodb://localhost:27017';

const mongoClient: MongoClient = await MongoClient.connect(url);

// get db instance
const db: Db = mongoClient.db('users');

// create the MongoDB Data provider creator function
const dataProviderCreator = createMongoDbProvider(db);

// highlight-end

// the business model
const userModel = `
 """
 @model
 """
 type User {
     id: ID!
     username: String!
 }
`;

// Use the dataProvider in buildGraphbackAPI
const { resolvers, services, contextCreator } = buildGraphbackAPI(userModel, { dataProviderCreator });
```

The highlighted code does the following:
 - Define connection URL to the MongoDB.
 - Create a connection to our Mongo database.
 - Select the `users` database.
 - And finally, create a data provider creator by using the `createMongoDbProvider` API. 
  
The rest of the code uses [`buildGraphbackAPI`](./api/graphback/modules/_buildgraphbackapi_.md) to create Graphback CRUD API based on the defined `userModel` model.

Visit [Data Models](../model/datamodel.md) to learn more about how to design your business models.
