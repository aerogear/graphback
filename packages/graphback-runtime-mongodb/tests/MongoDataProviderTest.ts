//tslint:disable-next-line: match-default-export-name
import { ObjectID } from 'mongodb';
import { advanceTo, advanceBy } from "jest-date-mock";
import { Context,createTestingContext } from "./__util__";

describe('MongoDBDataProvider Basic CRUD', () => {
  interface Todo {
    id: ObjectID;
    text: string;
  }
  let context: Context;

  const fields = ["id", "text"];

  const todoSchema = `
  """
  @model
  """
  type Todos {
  id: ID!
  text: String
  }
  `;

  const defaultTodoSeed = [
    {
      text: 'todo'
    },
    {
      text: 'todo2'
    }
  ]


  //Create a new database before each tests so that
  //all tests can run parallel


  afterEach(() => context.server.stop())

  test('Test mongo crud', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: defaultTodoSeed
      }
    });
    let todo: Todo = await context.providers.Todos.create({
      text: 'create a todo',
    }, {graphback: {services: {}, options: { selectedFields: fields}}});

    expect(todo.text).toEqual('create a todo');

    todo = await context.providers.Todos.update({
      id: todo.id,
      text: 'my updated first todo',
    }, {graphback: {services: {}, options: { selectedFields: fields}}});

    expect(todo.text).toEqual('my updated first todo');

    const data = await context.providers.Todos.delete({ id: todo.id }, {graphback: {services: {}, options: { selectedFields: fields}}});

    expect(data.id).toEqual(todo.id);
  });


  test('find first 1 todo(s) excluding first todo', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: defaultTodoSeed
      }
    });
    const todos = await context.providers.Todos.findBy({}, {graphback: {services: {}, options: { selectedFields: ["text"]}}}, undefined, { limit: 1, offset: 1 });

    expect(todos.length).toEqual(1);

    expect(todos[0].text).toEqual('todo2');
  });

  test('find Todo by text', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: defaultTodoSeed
      }
    });
    const all = await context.providers.Todos.findBy({}, {graphback: {services: {}, options: { selectedFields: []}}});
    const todos: Todo[] = await context.providers.Todos.findBy({
      text: { eq: all[0].text },
    }, {graphback: {services: {}, options: { selectedFields: ["id"]}}});
    expect(todos.length).toBeGreaterThan(0);
  });

  test('find Todo by text, limit defaults to complete set', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: defaultTodoSeed
      }
    });
    const text = 'todo-test';
    for (let i = 0; i < 11; i++) {
      await context.providers.Todos.create({
        text,
      }, {graphback: {services: {}, options: { selectedFields: fields}}});
    }
    const todos: Todo[] = await context.providers.Todos.findBy({ text: { eq: text } }, {graphback: {services: {}, options: { selectedFields: fields}}}, undefined, { offset: 0 });
    expect(todos.length).toEqual(11);
  });

  test('find by text offset defaults to 0', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: defaultTodoSeed
      }
    });
    const text = 'todo-test';
    for (let i = 0; i < 2; i++) {
      await context.providers.Todos.create({
        text,
      }, {graphback: {services: {}, options: { selectedFields: fields}}});
    }
    const todos: Todo[] = await context.providers.Todos.findBy({ text: { eq: text } }, {graphback: {services: {}, options: { selectedFields: fields}}},  undefined, { limit: 1 });
    expect(todos[0].text).toEqual(text);
  });

  test('find first 1 todos by text', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: defaultTodoSeed
      }
    });
    const text = 'todo-test';
    for (let i = 0; i < 2; i++) {
      await context.providers.Todos.create({
        text,
      }, {graphback: {services: {}, options: { selectedFields: fields}}});
    }

    const todos: Todo[] = await context.providers.Todos.findBy({ text: { eq: text } }, {graphback: {services: {}, options: { selectedFields: fields}}},  undefined, { limit: 1, offset: 0 });
    expect(todos.length).toEqual(1);
    expect(todos[0].text).toEqual(text);
  });

  test('test orderby w/o order', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: [
          { text: 'todo3' },
          { text: 'todo1' },
          { text: 'todo4' },
          { text: 'todo2' },
          { text: 'todo5' },
        ]
      }
    });

    const todos = await context.providers.Todos.findBy({ text: { contains: 'todo' } }, {graphback: {services: {}, options: { selectedFields: fields}}}, { field: 'text' });
    for (let t = 0; t < todos.length; t++) {
      expect(todos[t].text).toEqual(`todo${t + 1}`);
    }
  });

  test('test orderby with desc order', async () => {
    context = await createTestingContext(todoSchema, {
      seedData: {
        Todos: [
          { text: 'todo3' },
          { text: 'todo1' },
          { text: 'todo4' },
          { text: 'todo2' },
          { text: 'todo5' },
        ]
      }
    });

    const todos = await context.providers.Todos.findBy({ text: { contains: 'todo' } }, {graphback: {services: {}, options: { selectedFields: fields}}}, { field: 'text', order: 'desc' });
    for (let t = 0; t < todos.length; t++) {
      expect(todos[t].text).toEqual(`todo${5 - t}`);
    }
  });

  test('createdAt', async () => {
    context = await createTestingContext(`
    """
    @model
    @versioned
    """
    type Note {
      id: ID!
      text: String
    }
    `);
    const cDate = new Date(2020, 5, 26, 18, 29, 23);
    advanceTo(cDate);

    const res = await context.providers.Note.create({ text: 'asdf' }, {graphback: {services: {}, options: { selectedFields: ["id"]}}});
    expect(res.createdAt).toEqual(cDate.getTime());
    expect(res.createdAt).toEqual(res.updatedAt);
  })

  test('updatedAt', async () => {
    context = await createTestingContext(`
    """
    @model
    @versioned
    """
    type Note {
      id: ID!
      text: String
    }
    `);
    const createDate = new Date(2020, 5, 26, 18, 29, 23);
    advanceTo(createDate);

    const res = await context.providers.Note.create({ text: 'asdf' }, {graphback: {services: {}, options: { selectedFields: ["id"]}}});
    expect(res.updatedAt).toEqual(createDate.getTime());

    advanceBy(3000);
    const next = await context.providers.Note.update({
      ...res,
      text: 'asdftrains'
    }, {graphback: {services: {}, options: { selectedFields: ["updatedAt", "createdAt"]}}});

    const updateDate = new Date();
    expect(next.updatedAt).toEqual(updateDate.getTime());
    expect(next.createdAt).toEqual(createDate.getTime())
  })

  test('select only requested fields', async () => {
    context = await createTestingContext(`
    """
    @model
    """
    type Todos {
     id: ID!
     text: String,
     description: String
    }`, {
      seedData: {
        Todos: [
          { text: 'todo1', description: "first todo" },
          { text: 'todo2', description: "second todo" },
          { text: 'todo3', description: "third todo" }
        ]
      }
    });

    const todos = await context.providers.Todos.findBy({}, {graphback: {services: {}, options: { selectedFields: ["id", "text"]}}});

    expect(todos.length).toEqual(3);
    todos.forEach((todo) => {
      expect(todo.id).toBeDefined();
      expect(todo.text).toBeDefined();
      expect(todo.description).toBeUndefined(); // should be undefined since not selected
    });

    const createdTodo = await context.providers.Todos.create({text: "new todo", description: "todo add description"}, {graphback: {services: {}, options: { selectedFields: ["id"]}}});
    expect(createdTodo.id).toBeDefined();

    const updatedTodo = await context.providers.Todos.update({id: createdTodo.id, text: "updated todo"}, {graphback: {services: {}, options: { selectedFields: ["text"]}}});
    expect(updatedTodo.description).toBeUndefined();
    expect(updatedTodo.text).toEqual("updated todo");

    const deletedTodo = await context.providers.Todos.update({id: createdTodo.id}, {graphback: {services: {}, options: { selectedFields: ["id", "text", "description"]}}});
    expect(deletedTodo.id).toEqual(createdTodo.id);
    expect(deletedTodo.text).toEqual("updated todo");
    expect(deletedTodo.description).toEqual("todo add description");
  });

});
