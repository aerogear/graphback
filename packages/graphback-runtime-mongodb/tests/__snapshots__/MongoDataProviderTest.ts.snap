// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoDB indexing can create default indexes 1`] = `
Array [
  Object {
    "key": Object {
      "_id": 1,
    },
    "name": "_id_",
    "ns": "test.note",
    "v": 2,
  },
  Object {
    "key": Object {
      "text": 1,
    },
    "name": "text_1",
    "ns": "test.note",
    "v": 2,
  },
]
`;

exports[`MongoDB indexing can create indexes with options 1`] = `
Array [
  Object {
    "key": Object {
      "_id": 1,
    },
    "name": "_id_",
    "ns": "test.note",
    "v": 2,
  },
  Object {
    "key": Object {
      "meta": 1,
      "pages": 1,
    },
    "name": "compound_index",
    "ns": "test.note",
    "v": 2,
  },
]
`;
